lib/test_queue/iterator.rb

```diff
@@ -13,13 +13,22 @@ def initialize(sock, suites, filter=nil)
         @tcp_address = $1
         @tcp_port = $2.to_i
       end
+      @failures = 0
     end
 
     def each
       fail "already used this iterator. previous caller: #@done" if @done
 
       while true
-        client = connect_to_master('POP')
+        # If we've hit too many failures in one worker, assume the entire
+        # test suite is broken, and notify master so the run
+        # can be immediately halted.
+        if ENV["TEST_QUEUE_EARLY_FAILURE_LIMIT"] && @failures >= ENV["TEST_QUEUE_EARLY_FAILURE_LIMIT"].to_i
```

Comment:
If `TEST_QUEUE_EARLY_FAILURE_LIMIT` is static, is there a reason we'd want to read it from the ENV every time instead of storing it in a variable?

Also it'd be nice to validate `TEST_QUEUE_EARLY_FAILURE_LIMIT` is an int, otherwise we'll silently get string.to_i which is 0.